src/zcash/Proof.hpp

```diff
@@ -0,0 +1,211 @@
+#ifndef _ZCPROOF_H_
+#define _ZCPROOF_H_
+
+#include "serialize.h"
+#include "uint256.h"
+
+namespace libzcash {
+
+const unsigned char G1_PREFIX_MASK = 0x02;
+const unsigned char G2_PREFIX_MASK = 0x0a;
+
+// Element in the base field
+class Fq {
+private:
+    uint256 data;
+public:
+    Fq() : data() { }
+
+    template<typename libsnark_Fq>
+    Fq(libsnark_Fq element);
+
+    template<typename libsnark_Fq>
+    libsnark_Fq to_libsnark_fq() const;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(data);
+    }
+
+    friend bool operator==(const Fq& a, const Fq& b)
+    {
+        return (
+            a.data == b.data
+        );
+    }
+
+    friend bool operator!=(const Fq& a, const Fq& b)
+    {
+        return !(a == b);
+    }
+};
+
+// Compressed point in G1
+class CompressedG1 {
+private:
+    bool y_lsb;
+    Fq x;
+
+public:
+    CompressedG1() : y_lsb(false), x() { }
+
+    template<typename libsnark_G1>
+    CompressedG1(libsnark_G1 point);
+
+    template<typename libsnark_G1>
+    libsnark_G1 to_libsnark_g1() const;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        unsigned char leadingByte = G1_PREFIX_MASK;
+
+        if (y_lsb) {
+            leadingByte |= 1;
+        }
+
+        READWRITE(leadingByte);
+
+        if ((leadingByte & (~1)) != G1_PREFIX_MASK) {
+            throw std::ios_base::failure("lead byte of G1 point not recognized");
+        }
+
+        y_lsb = leadingByte & 1;
+
+        READWRITE(x);
+    }
+
+    friend bool operator==(const CompressedG1& a, const CompressedG1& b)
+    {
+        return (
+            a.y_lsb == b.y_lsb &&
+            a.x == b.x
+        );
+    }
+
+    friend bool operator!=(const CompressedG1& a, const CompressedG1& b)
+    {
+        return !(a == b);
+    }
+};
+
+// Compressed point in G2
+class CompressedG2 {
+private:
+    bool y_gt;
+    Fq x0;
+    Fq x1;
+
+public:
+    CompressedG2() : y_gt(false), x0(), x1() { }
+
+    template<typename libsnark_G2>
+    CompressedG2(libsnark_G2 point);
+
+    template<typename libsnark_G2>
+    libsnark_G2 to_libsnark_g2() const;
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        unsigned char leadingByte = G2_PREFIX_MASK;
+
+        if (y_gt) {
+            leadingByte |= 1;
+        }
+
+        READWRITE(leadingByte);
+
+        if ((leadingByte & (~1)) != G2_PREFIX_MASK) {
+            throw std::ios_base::failure("lead byte of G2 point not recognized");
+        }
+
+        y_gt = leadingByte & 1;
+
+        READWRITE(x0);
+        READWRITE(x1);
```

Comment:
x0 and x1 are being written in big endian already, it's just a matter of whether x0 or x1 goes first?